### ISSUE TITLE
Help with Website Agent

### ISSUE THREAD

**Issue by vx28643** [huginn/huginn#1216]
I want to email myself the full [APOD page](http://apod.nasa.gov/apod/astropix.html) whenever the site changes. But the full url of the site is not present in the img src, and the image does not display in email because the email client doesn't know where to look for it.

For example, http://apod.nasa.gov/apod/ap160109.html

in the source you will see...

<IMG SRC="image/1601/ngc1300_hst_960.jpg"

but to view in email I really need it to be:

<IMG SRC="http://apod.nasa.gov/apod/image/1601/ngc1300_hst_960.jpg"

So what's the simplest way to parse the text to ensure the image has the correct path?

I was hopeful to do this somehow with [XPath functions](http://www.w3.org/TR/xpath/#section-String-Functions) but couldn't find a way

I do want to email myself this entire page, although they also provide a RSS option.

**Comment by cantino** [comment]
Hey @vx28643.  Are you trying to email yourself the picture, or a message saying it's changed, or the full text of the page?

**Comment by vx28643** [comment]
Hey, thanks for your reply. In this case I'm trying to email myself the full page (html formatted text + image), in a way that the image will load in the email. The image should not be an attachment, but should load in the email just from html. 

I am currently using the following Website Agent to receive the full html of the page when it changes:

```
{
  "expected_update_period_in_days": "5",
  "url": "http://apod.nasa.gov/apod.rss",
  "type": "html",
  "mode": "on_change",
  "extract": {
    "content": {
      "css": "body",
      "value": "./node()"
    }
  }
}
```

But because the original website does not include it's domain in the IMG src, and only relative links to the image, the image is not loading in the email (see explanation in original question). 

Basically I want to add '"http://apod.nasa.gov/apod' to the beginning of the IMG, so that the location makes sense to the email app.

**Comment by vx28643** [comment]
oops made a mistake when first posting. corrected my reply above to "url": http://apod.nasa.gov/apod/

**Comment by cantino** [comment]
It's a bit ugly, but could you use an EventFormattingAgent with a Liquid tag like `{{ content | replace: 'image/', 'http://apod.nasa.gov/apod/image/' }}`?

**Comment by vx28643** [comment]
Brilliant. It worked! Thank you so much for your help.

**Comment by cantino** [comment]
Glad it worked!

**Comment by knu** [comment]
With #1217 merged, you could write something like this to resolve relative links in an HTML content:
```
{% regex_replace "(?<pre> (?:src|href)=(?<q>[\x22']))(?<href>.*)(?=\k<q>)" in %}{{ content }}{% with %}{{ pre }}{{ href | unescape | to_uri: "http://apod.nasa.gov/apod/" | replace_first: '' | escape }}{% endregex_replace %}
```
I know this is cryptic, but it should work.  I have a couple of regrets, though:

- The standard filter `escape` does not work against a non-string object (URI in this case) hence the `| replace_first: ''`part for stringification.
- In Ruby's regexp engine lookbehind does not work for a variable length pattern `(?:src|href)`, hence the `pre` capture group only to copy it.

**Comment by vx28643** [comment]
Hi @knu, that's fascinating and looks to be exactly what I want. However I'm having trouble understanding how to use it.

Below is my EventFormattingAgent with the replace tag @cantino suggested which works but obviously only works on links beginning with `image/`.
```
{
  "instructions": {
    "content": "{{ content | replace: 'image/', 'http://apod.nasa.gov/apod/image/' }}",
    "title": "{{title}}"
  },
  "matchers": [],
  "mode": "clean"
}
```
How would I modify this to use your `{% regex_replace...` code instead?

note: The http://apod.nasa.gov page contains an assortment of **href** and **img** tags that are to a mix of local and external pages. Ultimately I would like all links and images to resolve properly in the resulting html.

Regards.

**Comment by knu** [comment]
I think you can replace the "content" value (`{{ content | replace: ...}}`) with the said cryptic snippet and it should work.

**Comment by vx28643** [comment]
Does it need the quote-marks `"` escaped with  `\"` and/or anything else?

simply replacing with the snippet gives me:
```
"content": "{% regex_replace "(?<pre> (?:src|href)=(?<q>[\x22']))(?<href>.*)(?=\k<q>)" in %}{{ content }}{% with %}{{ pre }}{{ href | unescape | to_uri: "http://apod.nasa.gov/apod/" | replace_first: '' | escape }}{% endregex_replace %}"
```
and trying that gives me:
```
2 errors prohibited this Agent from being saved:
- Options was assigned invalid JSON
- instructions and mode need to be present.
```
I have then tried escaping the quotes and no JSON error, but it's not working thus far...

If you're not sure off the top of your head I will give it some more study and report back.

Thank you!

**Comment by vx28643** [comment]
Unfortunately I'm still not able to get it working. Perhaps I am not escaping the cryptic regex correctly. Here's the data I am using to test in EventFormattingAgent.

**EventFormattingAgent:**
```
{
  "instructions": {
    "content": "{% regex_replace \"(?<pre> (?:src|href)=(?<q>[x22']))(?<href>.*)(?=k<q>)\" in %}{{ content }}{% with %}{{ pre }}{{ href | unescape | to_uri: \"http://apod.nasa.gov/apod/\" | replace_first: '' | escape }}{% endregex_replace %}"
  },
  "matchers": [],
  "mode": "clean"
}
```
**Dry Run Data:**
```
{
  "content": "<a href=\"http:\/\/example.com\/test.html\">Good Link<\/a><a href=\"test.html\">Bad Link<\/a><IMG SRC=\"image\/test.jpg\">"
}
```

**Dry Run Results:**
```
[
  {
    "content": "<a href=\"http://example.com/test.html\">Good Link</a><a href=\"test.html\">Bad Link</a><IMG SRC=\"image/test.jpg\">"
  }
]
```

**Comment by knu** [comment]
The `\` right before `x22` is gone.  Doesn't just copy 'n paste on a JSON editor work?

**Comment by vx28643** [comment]
I have re-inserted the `\` right before the `x22` above (my mistake). I'm not sure which JSON editor to use. Pasting the cryptic string into http://www.jsoneditoronline.org/ does not auto-escape it. So far with Huginn I've just manually edited the JSON in Huginn itself.

Could you please fire up an EventFormattingAgent in Huginn and enter the agent data and Dry Run data from my example above? Perhaps you will find my error.

I really appreciate your continued assistance on this matter. Would love to get your regex working so it makes all links resolve.

**Comment by imkaihao** [comment]
@vx28643 
It's simpler to use the Liquid templating in the Website Agent, instead of regex replacing in the Event Formatting Agent.

If your Website Agent looks like this
```html
{
  "expected_update_period_in_days": "5",
  "url": "http://apod.nasa.gov/apod/",
  "type": "html",
  "mode": "on_change",
  "extract": {
    "content": {
      "css": "body",
      "value": "./node()"
    }
  }
}
```
You can change the agent to this
```html
{
  "expected_update_period_in_days": "5",
  "url": "http://apod.nasa.gov/apod/",
  "type": "html",
  "mode": "on_change",
  "extract": {
    "content": {
      "css": "body",
      "value": "./node()"
    }
  },
  "template": {
    "content_fixed": "{{ content | rebase_hrefs: _response_.url }}"
  }
}
```
The URLs in "content_fixed" become absolute URLs.

**Comment by rafdouglas** [comment]
Why not using the xpath "concatenate" function?
`
concat('http://baseurl.tld/',normalize-space(@href)) `

this Website Agent works smoothly, and does not require downsteam formatting agents:

```
{
  "expected_update_period_in_days": "5",
  "url": "http://apod.nasa.gov/apod/",
  "type": "html",
  "mode": "on_change",
  "extract": {
    "link": {
      "css": "img",
      "value": "concat('https://apod.nasa.gov/apod/',@src)"
    },
    "description": {
      "css": "center+ center b:nth-child(1)",
      "value": "normalize-space(.//text())"
    },
    "date": {
      "css": "center > p:nth-child(3)",
      "value": "normalize-space(.//text())"
    }
  }
}
```

**Comment by Garphy** [comment]
{
  "expected_update_period_in_days": "5",
  "url": "http://apod.nasa.gov/apod/",
  "type": "html",
  "mode": "on_change",
  "extract": {
    "content": {
      "css": "body",
      "value": "./node()"
    }
  },
  "template": {
    "content": "{{ content | to_uri: _response_.url }}"
  }
}

just like this.

**Comment by erebor** [comment]
@vx28643 Could this issue be closed?

**Comment by vx28643** [comment]
For certain, thanks very much for the help!
